natural_language_processing/word_embeddings/word_embeddings.py

```diff
@@ -0,0 +1,158 @@
+        assert all(self.is_in_vocab(x) for x in (x1, x2, y1)), error_msg
+        result = self._wv.most_similar(positive=[y1, x2], negative=[x1])[0][0]
+        return result
+
+    def n_similarity(self, list1: list, list2: list) -> float:
+        """
+        Returns similarity score between two lists of words
+        """
+        list1 = [word.lower() for word in list1 if self.is_in_vocab(word)]
+        list2 = [word.lower() for word in list2 if self.is_in_vocab(word)]
+        error_msg = 'at least one word from each list must be in the vocabulary'
+        assert list1 and list2, error_msg
+        score = self._wv.n_similarity(list1, list2)
+        return score
+
+    def similarity(self, w1: str, w2: str) -> float:
+        """
+        Returns similarity score between two words
+        """
+        w1, w2 = w1.lower(), w2.lower()
+        error_msg = 'both words must be in the vocabulary'
+        assert self.is_in_vocab(w1) and self.is_in_vocab(w2), error_msg
+        score = self._wv.similarity(w1, w2)
+        return score
+
+    def closest_words(self, word: str) -> list:
+        """
+        Returns similar words for a given word
+        """
+        word = word.lower()
+        error_msg = 'the word must be in the vocabulary'
+        assert self.is_in_vocab(word), error_msg
+        close_words = self._wv.similar_by_word(word)
+        return close_words
+
+    def plot_closest_words(self, word: str, plot=True):
+        """
+        plots similar words for a given word on a 2d space
+        """
+        close_words = self.closest_words(word)
+        words = [word for word, score in close_words]
+        words.append(word.lower())
+        vectors = self._wv[words]
+        tsne = TSNE(n_components=2)
+        X_tsne = tsne.fit_transform(vectors)
+        df = pd.DataFrame(X_tsne, index=words, columns=['x', 'y'])
+        fig = plt.figure()
+        ax = fig.add_subplot(1, 1, 1)
+        ax.scatter(df['x'], df['y'])
+        for word, pos in df.iterrows():
+            ax.annotate(word, pos)
+        if plot:
+            plt.show()
+        else:
+            fig.savefig('figure.png')
+
+    def is_in_vocab(self, word: str) -> bool:
+        """
+        Checks if model's vocabulary contains the word
+        """
+        return word.lower() in self._vocab
+
+    def save(self, filename='word2vec'):
+        """
+            Saves trained model
+        """
+        full_name = filename + '.model'
+        self._model.save(full_name)
+
+    @staticmethod
+    def _parse_document(text: str) -> list:
+        """
+        Parses give text document and returns list of tokens
+        """
+        tokens = word_tokenize(text)
+        splitted_tokens = []
+        for token in tokens:
+            splitted_tokens.extend(re.split('[^0-9a-zA-Z+.#]+', token))
+        parsed = [word.lower() for word in splitted_tokens
+                  if word and word not in stop_words]
+        return parsed
```

Comment:
```suggestion
        return [word.lower() for word in splitted_tokens
                if word and word not in stop_words]
```
The method name already tells us that these are parsed.